qiskit/circuit/quantumcircuit.py

```diff
@@ -499,6 +522,48 @@ def __eq__(self, other) -> bool:
             other, copy_operations=False
         )
 
+    def __copy__(self):
+        # Overridden to prevent copy.copy from using __{get,set}state__
+        # which currently unpacks _data into a Python list.
+        cls = self.__class__
+        cpy = cls.__new__(cls)
+        cpy.__dict__.update(self.__dict__)
+        return cpy
+
+    def __deepcopy__(self, memo=None):
+        # This is overridden in addition to __{get,set}state__
+        # to minimize memory pressure when we don't actually
+        # need to pickle (i.e. the typical deepcopy case).
+        cls = self.__class__
+        result = cls.__new__(cls)
+        for k in self.__dict__.keys() - {"_intern_context", "_data"}:
+            setattr(result, k, copy.deepcopy(self.__dict__[k], memo))
+
+        # Reuse the intern context to save space!
+        # TODO: there appears to be a concurrency issue when this
+        #   is done in a multi-threaded context (e.g. w/ primitives).
+        # result._intern_context = self._intern_context
+        result._intern_context = InternContext()
```

Comment:
This makes me very very nervous that we're missing important things, because this reuse is exactly _why_ the intern context is a separate object and not just a Rust-only handle within `CircuitData`.

This issue to me is a blocker on merge; I'd rather we don't realise the full performance due to an excess of synchronisation but are provably, _obviously_ correct when doing simple things like having multiple Python-space references to a single object.

I don't know what the problem is, but could we potentially skip the ref-counting within `InternContext`?  `QuantumCircuit` instances _shouldn't_ ever remove items from their data nor change the qubits/clbits an item acts on, so in principle, as long as a `QuantumCircuit` is alive, all of its interned things should be as well.  Sure, we'll keep around some memory for longer than we ought to in the case of chaining several `QuantumCircuit.copy_empty_like` calls together in between reconstructing entirely new circuits, but I don't think that'll be a huge effect (unless you have data otherwise?).  Certainly I'd _much_ rather keep the memory use a bit higher temporarily than to introduce race conditions into the data management.

Reply:
Addressed in a1e3f2b, which most importantly makes `InternContext` an internal Rust struct that we can still look into sharing in a future changeset. For now, we just clone it whenever we clone a `CircuitData`.

I'm quite certain the thread was being preempted (time slicing) while we had an active Rust borrow on the `PyCell` by a thread that also wanted to borrow. We don't expose `InternContext` to Python anymore (quite frankly, we don't have a current use case to anyway), so sharing the `InternContext` shouldn't be too difficult in the future.

Still, if we do decide to share it between `CircuitData` instances, we'll need some sort of concurrency control for accessing that shared mutable state. Naively, we could simply wrap it in an `RwLock`. But it'd be nice if there was a lock-free way instead since we rarely need concurrent access on the same `QuantumCircuit` instance.